tutorials/dataframe/df103_NanoAODHiggsAnalysis.h

```diff
@@ -0,0 +1,219 @@
+      z_masses[0] = el_z;
+      z_masses[1] = mu_z;
+   }
+   return z_masses;
+}
+
+// Compute Higgs mass from two electrons and two muons
+float compute_higgs_mass_2el2mu(rvec_f el_pt, rvec_f el_eta, rvec_f el_phi, rvec_f el_mass, rvec_f mu_pt, rvec_f mu_eta,
+                                rvec_f mu_phi, rvec_f mu_mass)
+{
+   TLorentzVector p1, p2, p3, p4;
+   p1.SetPtEtaPhiM(mu_pt[0], mu_eta[0], mu_phi[0], mu_mass[0]);
+   p2.SetPtEtaPhiM(mu_pt[1], mu_eta[1], mu_phi[1], mu_mass[1]);
+   p3.SetPtEtaPhiM(el_pt[0], el_eta[0], el_phi[0], el_mass[0]);
+   p4.SetPtEtaPhiM(el_pt[1], el_eta[1], el_phi[1], el_mass[1]);
+   return (p1 + p2 + p3 + p4).M();
+}
+
+// Compute Z masses from four leptons of the same kind and sort ascending in distance to Z mass
+RVec<float> compute_z_masses_4l(const RVec<RVec<size_t>> &idx, rvec_f pt, rvec_f eta, rvec_f phi, rvec_f mass)
+{
+   RVec<float> z_masses(2);
+   for (size_t i = 0; i < 2; i++) {
+      TLorentzVector p1, p2;
+      const auto i1 = idx[i][0]; const auto i2 = idx[i][1];
+      p1.SetPtEtaPhiM(pt[i1], eta[i1], phi[i1], mass[i1]);
+      p2.SetPtEtaPhiM(pt[i2], eta[i2], phi[i2], mass[i2]);
+      z_masses[i] = (p1 + p2).M();
+   }
+   if (std::abs(z_masses[0] - z_mass) < std::abs(z_masses[1] - z_mass)) {
+      return z_masses;
+   } else {
+      return Reverse(z_masses);
+   }
+}
+
+// Compute mass of Higgs from four leptons of the same kind
+float compute_higgs_mass_4l(const RVec<RVec<size_t>> &idx, rvec_f pt, rvec_f eta, rvec_f phi, rvec_f mass)
+{
+   TLorentzVector p1, p2, p3, p4;
+   const auto i1 = idx[0][0]; const auto i2 = idx[0][1];
+   const auto i3 = idx[1][0]; const auto i4 = idx[1][1];
+   p1.SetPtEtaPhiM(pt[i1], eta[i1], phi[i1], mass[i1]);
+   p2.SetPtEtaPhiM(pt[i2], eta[i2], phi[i2], mass[i2]);
+   p3.SetPtEtaPhiM(pt[i3], eta[i3], phi[i3], mass[i3]);
+   p4.SetPtEtaPhiM(pt[i4], eta[i4], phi[i4], mass[i4]);
+   return (p1 + p2 + p3 + p4).M();
+}
+
+bool filter_z_dr(const RVec<RVec<size_t>> &idx, rvec_f eta, rvec_f phi)
+{
+   for (size_t i = 0; i < 2; i++) {
+      const auto i1 = idx[i][0];
+      const auto i2 = idx[i][1];
+      const auto dr = sqrt(pow(eta[i1] - eta[i2], 2) + pow(phi[i1] - phi[i2], 2));
+      if (dr < 0.02) {
+         return false;
+      }
+   }
+   return true;
+};
+
+bool pt_cuts(rvec_f mu_pt, rvec_f el_pt)
+{
+   auto mu_pt_sorted = Reverse(Sort(mu_pt));
+   if (mu_pt_sorted[0] > 20 && mu_pt_sorted[1] > 10) {
+      return true;
+   }
+   auto el_pt_sorted = Reverse(Sort(el_pt));
+   if (el_pt_sorted[0] > 20 && el_pt_sorted[1] > 10) {
+      return true;
+   }
+   return false;
+}
+
+bool dr_cuts(rvec_f mu_eta, rvec_f mu_phi, rvec_f el_eta, rvec_f el_phi)
+{
+   auto mu_dr = sqrt(pow(mu_eta[0] - mu_eta[1], 2) + pow(mu_phi[0] - mu_phi[1], 2));
+   auto el_dr = sqrt(pow(el_eta[0] - el_eta[1], 2) + pow(el_phi[0] - el_phi[1], 2));
+   if (mu_dr < 0.02 || el_dr < 0.02) {
+      return false;
+   }
+   return true;
+}
+
+float add_weight_higgs_sig_4mu()
```

Comment:
why are these functions call `add_*`? they don't add things..?

Reply:
yeah well I may have been too quick in defining these functions, thinking they "add" the weight columns to the dataframe... Maybe a better name would be `define_weight_column_*` ?

Reply:
I'd say just `weight_*` or `eval_weight_*`.

Also, for all these functions that just return constants it's probably better to just do something like  `df.Define("column", f"{constant}")` than invoke the function at every entry (one day we'll have `DefineConstant`...)